### PR TITLE
Fix interop between Python 3 master and Python 2 worker

### LINKED ISSUE #3416: python 3 master <==> python 2 worker, does not work
From the buildbot master branch, I experimented with setting up
a Python 3.6 master and a Python 2.7 worker.  I was able to start the master and worker,
and saw that they connected to each other.

However, when I triggered a build, I got:

```
Traceback (most recent call last):
  File "/home/vagrant/buildbot-venv/lib64/python3.6/site-packages/twisted/internet/defer.py", line 653, in _runCallbacks
    current.result = callback(current.result, *args, **kw)
  File "/home/vagrant/buildbot-venv/lib64/python3.6/site-packages/twisted/internet/defer.py", line 1442, in gotResult
    _inlineCallbacks(r, g, deferred)
  File "/home/vagrant/buildbot-venv/lib64/python3.6/site-packages/twisted/internet/defer.py", line 1384, in _inlineCallbacks
    result = result.throwExceptionIntoGenerator(g)
  File "/home/vagrant/buildbot-venv/lib64/python3.6/site-packages/twisted/python/failure.py", line 393, in throwExceptionIntoGenerator
    return g.throw(self.type, self.value, self.tb)
--- <exception caught here> ---
  File "/home/vagrant/buildbot/master/buildbot/process/buildstep.py", line 574, in startStep
    self.results = yield self.run()
  File "/home/vagrant/buildbot-venv/lib64/python3.6/site-packages/twisted/internet/defer.py", line 1386, in _inlineCallbacks
    result = g.send(result)
  File "/home/vagrant/buildbot/master/buildbot/process/buildstep.py", line 704, in run
    results = yield self._start_deferred
  File "/home/vagrant/buildbot/master/buildbot/steps/source/git.py", line 171, in startVC
    gitInstalled = yield self.checkBranchSupport()
  File "/home/vagrant/buildbot-venv/lib64/python3.6/site-packages/twisted/internet/defer.py", line 1384, in _inlineCallbacks
    result = result.throwExceptionIntoGenerator(g)
  File "/home/vagrant/buildbot-venv/lib64/python3.6/site-packages/twisted/python/failure.py", line 393, in throwExceptionIntoGenerator
    return g.throw(self.type, self.value, self.tb)
  File "/home/vagrant/buildbot/master/buildbot/steps/source/git.py", line 583, in checkBranchSupport
    stdout = yield self._dovccmd(['--version'], collectStdout=True)
  File "/home/vagrant/buildbot-venv/lib64/python3.6/site-packages/twisted/internet/defer.py", line 1384, in _inlineCallbacks
    result = result.throwExceptionIntoGenerator(g)
  File "/home/vagrant/buildbot-venv/lib64/python3.6/site-packages/twisted/python/failure.py", line 393, in throwExceptionIntoGenerator
    return g.throw(self.type, self.value, self.tb)
  File "/home/vagrant/buildbot/master/buildbot/steps/source/git.py", line 381, in _dovccmd
    yield self.runCommand(cmd)
  File "/home/vagrant/buildbot-venv/lib64/python3.6/site-packages/twisted/internet/defer.py", line 1384, in _inlineCallbacks
    result = result.throwExceptionIntoGenerator(g)
  File "/home/vagrant/buildbot-venv/lib64/python3.6/site-packages/twisted/python/failure.py", line 393, in throwExceptionIntoGenerator
    return g.throw(self.type, self.value, self.tb)
  File "/home/vagrant/buildbot/master/buildbot/process/buildstep.py", line 900, in runCommand
    res = yield command.run(self, self.remote, self.build.builder.name)
buildbot.process.remotecommand.RemoteException: b'exceptions.AssertionError': b"Unexpected usePTY argument value: u'slave-config'. Expected boolean."
b'Traceback (most recent call last):\n  File "/home/vagrant/buildbot-venv-2.7/lib/python2.7/site-packages/twisted/spread/pb.py", line 1054, in _recvMessage\n    netResult = object.remoteMessageReceived(self, message, netArgs, netKw)\n  File "/home/vagrant/buildbot-venv-2.7/lib/python2.7/site-packages/twisted/spread/flavors.py", line 125, in remoteMessageReceived\n    state = method(*args, **kw)\n  File "/home/vagrant/buildbot/worker/buildbot_worker/base.py", line 150, in remote_startCommand\n    d = self.command.doStart()\n  File "/home/vagrant/buildbot/worker/buildbot_worker/commands/base.py", line 163, in doStart\n    d = defer.maybeDeferred(self.start)\n--- <exception caught here> ---\n  File "/home/vagrant/buildbot-venv-2.7/lib/python2.7/site-packages/twisted/internet/defer.py", line 150, in maybeDeferred\n    result = f(*args, **kw)\n  File "/home/vagrant/buildbot/worker/buildbot_worker/commands/shell.py", line 47, in start\n    logEnviron=args.get(\'logEnviron\', True),\n  File "/home/vagrant/buildbot/worker/buildbot_worker/runprocess.py", line 388, in __init__\n    usePTY)\nexceptions.AssertionError: Unexpected usePTY argument value: u\'slave-config\'. Expected boolean.\n'
```

### PR DESCRIPTION
**Pull request by rodrigc**
Fixes #3416

### CODE REVIEW

**Comment by seankelly** [comment]
Why the change from using native string to unicode on the worker?

**Comment by rodrigc** [comment]
@seankelly I was never happy with `bytes2NativeString` on the worker.  However, it was a crutch which allowed me to move forward with Python 3 porting, and keep the worker behaving the same way on Python 2.  Now that all the tests pass on Python 3, I can take a deeper look at everything involved, and try to take a pass at solving some more difficult scenarios and corner cases.

It took me a long time of debugging Python 3 master <==> Python 2 worker interop to figure out what was going on.  I decided to switch the worker to internally use unicode so that the behavior was the same on Python 2 and 3.  This made things a bit easier to figure out, otherwise my mind was doing contortions trying to figure out the permutations of what was going on in Python 3 and Python 2 on the worker.

**Comment by rodrigc** [comment]
With this patch, in my initial testing, I was able to have a Python 3 master, and a Python 2 worker, and have them successfully interoperate, i.e. execute build steps without the errors seen in #3416

**Comment by rene00** [comment]
I'm running 3.6 master and a 2.7 slave. This branch is working well for me so far. https://github.com/buildbot/buildbot/issues/3416 doesnt occur.

**Comment by rodrigc** [comment]
@congzhangzh you reported a problem in https://github.com/buildbot/buildbot/issues/3631
Would you be able to try the patch in this PR on your system?

I made a lot of changes to how bytes/unicode is handled in the buildbot worker in this patch.

You can take the files I changed directly from this PR, or you can install my branch directly using
pip inside a virtualenv:

```
python -m pip install --upgrade "git+https://github.com/rodrigc/buildbot@py2-interop#egg=buildbot&subdirectory=master"
python -m pip install --upgrade "git+https://github.com/rodrigc/buildbot@py2-interop#egg=buildbot-worker&subdirectory=worker"
python -m pip install buildbot-www buildbot-grid-view buildbot-waterfall-view buildbot-console-view
```

**Comment by rodrigc** [comment]
@rene00 wow thanks for providing feedback so fast.  Since I've got your attention,
do you think you can run a quick test with this branch, and try to run a Python 2 master with a Python 3 worker?

**Comment by tardyp** [comment]
@rodrigc thanks for working on this, and pushing tw 17.9. That will be a great milestone for 0.9.12!

I am okay to push for having all the exchange made in unicode. 

However this seams incoherent with what you say earlier, were it is safest to just put bytes on the wire everywhere, and transform to unicode as soon as possible when manipulating in python.
I had a concern about performance on the latter method, as I think we currently have performance and memory bloat issues, because we are converting back and forth several time every line of logs received from the workers. Logs come from worker process as (locale or configured encoding) bytes, then are converted in worker to unicode, then put to wire as bytes, then put to master in unicode, and then in db as utf8 bytes
This is something that we will have to eventually optimize.

I am okay to do the functionality first, and then optimise, but I think we definitively need those integration tests to make sure this py2/py3 compat does not regress.

**Comment by rodrigc** [comment]
@rene00 I just pushed some more fixes to this branch, and now I was able to successfully test:
Python 2 master <==> Python 3 worker

File: worker/buildbot_worker/pbutil.py [hunk 1]
```diff
@@ -123,3 +126,19 @@ def failedToGetPerspective(self, why, broker):
 
         # lose the current connection, which will trigger a retry
         broker.transport.loseConnection()
+
+
+def decode(data, encoding='utf-8', errors='strict'):
+    """We need to convert a dictionary where keys and values
+    are bytes, to unicode strings.  This happens when a
+    Python 2 worker sends a dictionary back to a Python 3 master.
```
Comment (tardyp) [inline worker/buildbot_worker/pbutil.py]: Python2 master sends a dictionarry to python3 worker?

**Comment by tardyp** [comment]
looks good apart from this copy paste mistake

File: worker/buildbot_worker/pbutil.py (hunk 1 above)
Comment (rodrigc) [reply worker/buildbot_worker/pbutil.py]: Fixed

**Comment by rodrigc** [comment]
@tardyp my first pass on this is to get py2 <==> py3 interop working at all.
I agree there could be further improvements.

**Comment by rodrigc** [comment]
I think the best further improvement would be to extend the worker so
that it passes back the `unicode_encoding` used on the worker to the master,
and passes the log as bytes back to the worker.
